modules/doctor/repository.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package doctor
+
+import (
+	"context"
+
+	"code.gitea.io/gitea/models/db"
+	user_model "code.gitea.io/gitea/models/user"
+	repo_service "code.gitea.io/gitea/services/repository"
+
+	"xorm.io/builder"
+)
+
+// CountOrphanedRepos count repository where user of owner_id do not exist
+func CountOrphanedRepos(ctx context.Context) (int64, error) {
+	return db.CountOrphanedObjects(ctx, "repository", "user", "repository.owner_id=user.id")
+}
+
+// DeleteOrphanedRepos delete repository where user of owner_id do not exist
+func DeleteOrphanedRepos(ctx context.Context) (int64, error) {
+	batchSize := db.MaxBatchInsertSize("repository")
+	e := db.GetEngine(ctx)
+	var deleted int64
+	adminUser := &user_model.User{IsAdmin: true}
+
+	for {
+		var ids []int64
+		if err := e.Table("`repository`").
+			Join("LEFT", "`user`", "repository.owner_id=user.id").
+			Where(builder.IsNull{"`user`.id"}).
+			Select("`repository`.id").Limit(batchSize).Find(&ids); err != nil {
+			return deleted, err
+		}
+
+		// if we don't get ids we have deleted them all
+		if len(ids) == 0 {
+			return deleted, nil
+		}
+
+		for _, id := range ids {
+			if err := repo_service.DeleteRepositoryDirectly(ctx, adminUser, 0, id, true); err != nil {
```

Comment:
Why not remove repo_id from team_repo table?

Reply:
Because that record is already successfully deleted

Reply:
-> https://github.com/go-gitea/gitea/pull/27290#issuecomment-1736524793

Reply:
> Because that record is already successfully deleted

Where? Sorry, I still don't understand how it's realted with https://github.com/go-gitea/gitea/pull/27290#issuecomment-1736524793

Reply:
Gitea has a bug where the user and it's orgs get as intendet sucessfully deleted. But the repos of that org are not

Reply:
I think we can delete it again even if all of them have been deleted or there is no any record about this repository on team_repo.

Reply:
as the user do not exist anymore we dont know it it actually was an or or not

what I could do is to add another check that deletes team_repo's independent if there is no corresponding repo.

I could also rewrite a dedicated deletion function but I want to avoid code duplication